Add simple 1D reservoir model

mikekryjak opened the pull request:
A major limitation of 1D SOL models is the inability to consider cross-field transport. This is particularly impactful in deep detachment where the neutral pressure can be many times higher than in a comparable 2D model where neutrals can spread across field lines and get pumped out.

This PR adds a simple reservoir model based on the one in DIV1D but a step simpler. The user defines the reservoir location by passing an expression to `reservoir_location` and sets it to a constant density with `reservoir_density`. The exchange term for density will be proportional to the difference between the domain density and the reservoir density, and there will be proportional pressure and momentum sinks. In regions where the particle transfer goes in the other direction, new particles will appear with the same pressure and momentum as the local ones.

Still trying to work out exactly what DIV1D did but I think they evolve the reservoir density with a domain exchange term as described above plus a pump sink term. This is according to Gijs Derks' MSc thesis and may have changed since then.

The reservoir can be implemented for any species.

- [x] Implement component
- [x] Implement exchange channels
- [ ] Think about whether we need to evolve density
- [ ] Think about what pressure and momentum the returning particles should be
- [ ] Fix description of energy source (currently pressure)
- [ ] Test
- [ ] Docs


If we have evolving reservoirs we need to evolve a scalar. Peter prepared an example: https://github.com/boutproject/BOUT-dev/blob/evolve-scalars/tests/integrated/test-solve-scalar/test_solve_scalar.cxx

mikekryjak committed:
Add reservoir component and reservoir location

mikekryjak committed:
Fix passing "name" to options and reservoir location type

mikekryjak committed:
Complete reservoir implementation

mikekryjak committed:
Add diagnostic prints and change the set source to add

bendudson commented on src/reservoir.cxx:
```diff
@@ -0,0 +1,220 @@
+  AUTO_TRACE();
+
+  // We are operating on only one species
+  auto species = state["species"][name];
```
```suggestion
  auto& species = state["species"][name];
```

bendudson replied:
@ZedThree question: Is there a way that we can force copies to be explicit? Perhaps delete the copy constructor and add a `copy` function so that 
```
auto species = state["species"][name];
```
would be a compile error. If someone really wants to copy they have to make it explicit:
```
auto species = copy(state["species"][name]);
```
Having the behavior depend on a single `&` in the right place is a footgun that keeps going off.

bendudson commented on src/reservoir.cxx:
```diff
@@ -0,0 +1,220 @@
+      for(int iy=0; iy < mesh->LocalNy ; iy++){
+          for(int iz=0; iz < mesh->LocalNz; iz++){
+
+            auto i = indexAt(N, ix, iy, iz);
```
I think this triple-nested loop and `indexAt` can be replaced by:
```
BOUT_FOR(i, N.getRegion("RGN_NOBNDRY")) {
   ...
}
```

bendudson committed:
Use reference to Options

Accidentally making a copy of the state, rather than getting a reference.

bendudson committed:
reservoir: Tidying code, adding comments

- Deleted commented-out code
- Replaced nested loop with BOUT_FOR
- Use GET_NOBOUNDARY so that more useful errors are printed
  if values are modified out of order
- Add comments to class, moving some from .cxx to header
- Delete unused variables
- Run clang-format

bendudson committed:
reservoir: Create reservoir_region to simplify iteration

Rather than checking `reservoir_region` every time `transform` is
called, create a `Region` containing only the cell indices needed.

mikekryjak committed:
Add flag to prevent reservoir from adding particles

mikekryjak commented:
There is an issue where the final domain cell is seeing an extremely high density source in the last domain cell. Here is a plot near the target which includes guard cells showing the model output and a post-processing calculation which is identical to what's in the code.

![image](https://github.com/bendudson/hermes-3/assets/62797494/933af049-48a5-45d1-b5d3-fdc5076f4447)

There is no large energy source there which is bizarre since it should be proportional. However, there is a large pressure source in the first guard cell for whatever reason.
![image](https://github.com/bendudson/hermes-3/assets/62797494/b0847efe-b008-496d-8657-0bbfdf2f4d49)

The momentum source matches the calculation, unlike energy and density. Again, this is very strange since the momentum source is proportional to the density source:
![image](https://github.com/bendudson/hermes-3/assets/62797494/1b088048-b21a-4622-9572-6123168bebd6)

No clue what's going on here. I think I will start by setting the guard cells to 0 in case they are having an impact, but I thought they wouldn't. @bendudson any ideas?

mikekryjak committed:
Merge branch 'master' into reservoir-model

mikekryjak committed:
Add ypos calculation and inputs for xpoint/baffle pos

mikekryjak committed:
Change ypos to lpar

mikekryjak committed:
Extend to 3 reservoirs as per Derks 2024

mikekryjak committed:
Take out unnecessary prints

mikekryjak committed:
Merge branch 'master' into reservoir-model
